node/external/transactionAPI/apiTransactionResults.go

```diff
@@ -0,0 +1,222 @@
+package transactionAPI
+
+import (
+	"encoding/hex"
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/data/receipt"
+	"github.com/ElrondNetwork/elrond-go-core/data/smartContractResult"
+	"github.com/ElrondNetwork/elrond-go-core/data/transaction"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/dblookupext"
+	"github.com/ElrondNetwork/elrond-go/node/filters"
+)
+
+type apiTransactionResultsProcessor struct {
+	addressPubKeyConverter core.PubkeyConverter
+	historyRepository      dblookupext.HistoryRepository
+	storageService         dataRetriever.StorageService
+	marshalizer            marshal.Marshalizer
+	selfShardID            uint32
+}
+
+func newAPITransactionResultProcessor(
+	addressPubKeyConverter core.PubkeyConverter,
+	historyRepository dblookupext.HistoryRepository,
+	storageService dataRetriever.StorageService,
+	marshalizer marshal.Marshalizer,
+	selfShardID uint32,
+) *apiTransactionResultsProcessor {
+	return &apiTransactionResultsProcessor{
+		addressPubKeyConverter: addressPubKeyConverter,
+		historyRepository:      historyRepository,
+		storageService:         storageService,
+		marshalizer:            marshalizer,
+		selfShardID:            selfShardID,
+	}
+}
+
+func (arp *apiTransactionResultsProcessor) putResultsInTransaction(hash []byte, tx *transaction.ApiTransactionResult, epoch uint32) {
+	arp.putLogsInTransaction(hash, tx, epoch)
+
+	resultsHashes, err := arp.historyRepository.GetResultsHashesByTxHash(hash, epoch)
+	if err != nil {
+		return
+	}
+
+	if len(resultsHashes.ReceiptsHash) > 0 {
+		arp.putReceiptInTransaction(tx, resultsHashes.ReceiptsHash, epoch)
+		return
+	}
+
+	arp.putSmartContractResultsInTransaction(tx, resultsHashes.ScResultsHashesAndEpoch)
+}
+
+func (arp *apiTransactionResultsProcessor) putReceiptInTransaction(tx *transaction.ApiTransactionResult, recHash []byte, epoch uint32) {
+	rec, err := arp.getReceiptFromStorage(recHash, epoch)
+	if err != nil {
+		log.Warn("nodeTransactionEvents.putReceiptInTransaction() cannot get receipt from storage",
+			"hash", hex.EncodeToString(recHash))
+		return
+	}
+
+	tx.Receipt = arp.adaptReceipt(rec)
+}
+
+func (arp *apiTransactionResultsProcessor) getReceiptFromStorage(hash []byte, epoch uint32) (*receipt.Receipt, error) {
+	receiptsStorer := arp.storageService.GetStorer(dataRetriever.UnsignedTransactionUnit)
+	receiptBytes, err := receiptsStorer.GetFromEpoch(hash, epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	rec := &receipt.Receipt{}
```

Comment:
you can reduce the duplicate code by using the `unmarshalerAndPreparer` implementation

Reply:
refactored.